engine.go

```
@@ -142,13 +199,168 @@ func (e *Engine) RunSync(ctx context.Context, w *Workflow) (err error) {
+		}
+
+		w.CurrentStep = i + 1
+		if e.store != nil {
+			if err := e.store.SaveWorkflow(ctx, w); err != nil {
+				return fmt.Errorf("failed to save workflow state after step %d: %v", w.CurrentStep-1, err)
+			}
+		}
+	}
+
+	w.Status = StatusCompleted
+	if e.store != nil {
+		if err := e.store.SaveWorkflow(ctx, w); err != nil {
+			return fmt.Errorf("failed to save final workflow state: %w", err)
+		}
+	}
+	return nil
+}
+
+func (e *Engine) runWithQueue(ctx context.Context, w *Workflow, queueName string) {
+	defaultQueueOptions := QueueOptions{
+		AutoDelete:  true,
+		DeleteAfter: 1 * time.Minute,
+		PopTimeout:  1 * time.Second,
+	}
+
+	defaultWorkersDef := WorkersDefinition{
+		Count:        1,
+		PollInterval: 1 * time.Second,
+		WorkTimeout:  5 * time.Minute}
+
+	err := e.CreateQueue(ctx, queueName, defaultWorkersDef, defaultQueueOptions)
+	if err != nil && err != ErrQueueAlreadyExists {
+		log.Printf("failed to create queue %s: %v", queueName, err)
+		return
+	}
+
+	// in case of ErrQueueAlreadyExists or queue is just created, we need to get the queue
+	queue, err := e.queueEngine.GetQueue(ctx, queueName)
+	if err != nil {
+		log.Printf("failed to get queue %s: %v", queueName, err)
+		return
+	}
+
+	// enqueue workflow
+	if err := queue.Enqueue(ctx, w); err != nil {
+		log.Printf("failed to enqueue workflow %s to queue %s: %v", w.Name, queueName, err)
+		return
+	}
 }
 
 // RunAsync runs a workflow asynchronously in a new goroutine.
 // Errors are logged to standard output.
 // RunAsync runs a workflow asynchronously in a new goroutine.
-func (e *Engine) RunAsync(ctx context.Context, w *Workflow) {
+// If a queue name is provided, it enqueues the workflow instead of running it directly.
+func (e *Engine) RunAsync(ctx context.Context, w *Workflow, opts ...RunOption) {
+	options := &runOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+	
+	if options.queueName != "" {
+		e.runWithQueue(ctx, w, options.queueName)
```

I think we should return an error if we didn't manage to schedule the workflow